Pull in new enemy, enemy AI, enemy spawner, and health system implementations

edjohn opened the pull request:
Add general health system and player health scripts; Add enemy AI finite state machine scripts; Add enemy script with animation handler and associated animator (currently buggy and unimplemented); Add enemy spawner script; Add prefab for striker enemy with new AI and enemy scripts; Add laser material for weapon targeting

Commit:
Add general health system and player health scripts; Add enemy AI finite state machine scripts; Add enemy script with animation handler and associated animator (currently buggy and unimplemented); Add enemy spawner; Add prefab for striker enemy with new AI and enemy scripts; Add laser material for weapon targeting

dragon2601 committed:
User interface (#23)

* Working on the new menu

#17

* working on the main menu, added sound and fading features

#17

* Main Menu built

#17

* font changed for main menu

#17

* fixed bugs(screen resolution)

#17

JaredGivens commented on Assets/Scripts/Enemy Ai/AttackState.cs:
```diff
@@ -16,7 +16,11 @@ public class AttackState : State
         //player = GameObject.Find("TopDownCharacter");
+    }
 
+    public override string GetStateAsString()
```
This looks really bad. Override implies that this is a virtual function call that incurs 3 cache misses, and it is very bad for performance. Additionally, we should try to avoid introducing a string period. There's no reason to use a dynamically sized char array to identify a state. Ideally, we would use an Enum or typeof.

abhiss replied:
Technically strings in c# are immutable and statically sized. Using strings to identify the type of the state is bad though.

abhiss commented on Assets/Scripts/Character/HealthSystem.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    private void OnTriggerEnter(Collider other)
```
why is the health system doing damage? shouldn't this happen in the attack state?

abhiss commented on Assets/Scripts/Character/PlayerHealth.cs:
```diff
@@ -0,0 +1,49 @@
+using UnityEngine;
+
+public class PlayerHealthSystem : MonoBehaviour
```
It looks like HealthSystem.cs and PlayerHealth.cs do the same thing. I think it should be abstracted into a single class that can be put on players and enemies.

abhiss commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -0,0 +1,84 @@
+    private EnemyAnimationHandler _animationHandler;
+    */
+
+    public float Health
```
these should be auto properties https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties

abhiss commented on Assets/Scripts/Enemy/EnemyAnimationHandler.cs:
```diff
@@ -0,0 +1,46 @@
+            _animator.SetTrigger("Die");
+        }
+        //  Handle enemy movement animations based on current state.
+        switch (_currentState)
```
all this abstraction to switch over a string? :(

abhiss reviewed (changes requested):
1. i think GetCurrentStateAsString is a bad idea and we should use typeof or an enum or something
2. HealthSystem and PlayerHealth should be a single component

dominicaq commented:
I'm okay with the abstractions as long as they don't hamper your ability to create and modify enemies in a reasonable time frame. If the abstractions are a noticeable roadblock in development I suggest you start from scratch before we get any closer to the deadline. I highly recommend you learn about coroutines, as I believe they will be immensely helpful in creating your state machine logic. Remember, we're not developing a game for 1-2 years. We have 4 weeks and are students with homework and exams from other classes. We're making at most 3-4 unique enemies and a boss, keeping the code simple will allow you to quickly modify the code without having to worry about the overall structure.

dominicaq commented on Assets/Scripts/Character/HealthSystem.cs:
```diff
@@ -0,0 +1,52 @@
+    private void Die()
+    {
+        // Implement any logic needed when the object dies, e.g., destroying the object
+        if (gameObject.CompareTag("TopDownCharacter"))
```
The tag manager has a “Player” tag already, you don’t need to create a new one. A good rule of thumb is when you’re considering tags there’s better alternatives aside from dealing with the OnCollision functions from mono behavior. That being said if you absolutely had to use tags, CompareTag is the way to go as it compiles into fewer instructions than myTag == tag. See https://answers.unity.com/questions/200820/is-comparetag-better-than-gameobjecttag-performanc.html

dominicaq commented on Assets/Scripts/Character/PlayerHealth.cs:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    void OnTriggerStay(Collider other)
+    {
```
I get what you’re doing here but your naming is a bit confusing. Damage over time should be a different class based on the context. You’re trying to do invincibility frames here but it’s not clear when reading at a glance. This code shouldn’t be in OnTrigger, but should be in TakeDamage. You would just return before sending damage to the health bar with the logic you have here minus the CompareTag.  If your code is not doing this then it should not be in this script.

abhiss commented:
fyi dominics branch added the health system that jackie made earlier and this branch has another health system so that's gonna be a conflict. we'll probably keep the other one since it already works with shooting now.

dominicaq commented on Assets/Scripts/Enemy/EnemySpawner.cs:
Sorry for the late comment. Please discard this file or heavily modify it to fit the needs of the project. The code present does not represent the proposed spawn system we have in mind. Our spawn system should do the following:

Dynamic Spawning:
- Check if its a valid spawn (No spawning in the void)
- Spawn in waves off-screen from **ALL** player's screens.
- Ease of controlling spawn rates so other scripts may control the game flow

We will also have baked spawning, which will be handled by Jared.

Also, no function overloading in this instance where you're calling a single function
```
Spawn(nextEnemyToSpawn, spawnPos);
```

Just call the already abstracted instantiate function instead.

edjohn commented:
![crying-emoji-dies](https://github.com/abhiss/symmetrical-octo-waffle/assets/54566278/56a4f50b-6172-4f95-ab01-85c9c9893df1)

this pr aint making it out the hood

Commit:
Import tit
an model; Refactor animation handling to type switches and modify speed during a
nimations; Correctly disable navMeshAgent when enemy is dead; Add titan enemy wi
th animations; Remove GetCurrentStateAsString() from enemy states

Commit:
Add titan enemy prefab missed in previous commit

Commit:
Add state gameobjects to titan prefab

Commit:
Remove unnecssary code from Enemy.cs; Add documentation to EnemyAnimationHandler and Enemy.cs; Fix Titan prefab not chasing correctly by setting state positions to (0,0,0)

Commit:
Fix titan not playing death animation while mid-attack

Commit:
Merge main; Remove Jackie health scripts; Give Titan enemy prefab correct health system and update Enemy.cs to work with correct health system; Re-add titan model

Commit:
Refactor Enemy and EnemyAnimationHandler; Implement a death coroutine for enemy

Commit:
Fix animation controller so attack animations can be played consistently
